### ISSUE TITLE
[BUG] Performance regression for crossjoin on 0.6 comparing to 0.5

### ISSUE THREAD

**Issue by viadea** [NVIDIA/spark-rapids#2371]
**Describe the bug**

For a 2-million rows cross-join, it completes in 14mins on 0.6 snapshot while it can complete in around 5+ mins on 0.5 snapshot.
This is done using single node spark standalone cluster based on RTX6000.

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.

The `customer` table has 2 million rows.
On 0.5:
```
spark.conf.set("spark.rapids.sql.exec.CartesianProductExec",false)
spark.conf.set("spark.rapids.sql.exec.BroadcastNestedLoopJoinExec",true)
spark.conf.set("spark.sql.autoBroadcastJoinThreshold",1000000000)
spark.conf.set("spark.sql.shuffle.partitions",200)
spark.conf.set("spark.rapids.sql.batchSizeBytes",1048576)
spark.conf.set("spark.rapids.sql.reader.batchSizeRows",1048576)
spark.conf.set("spark.sql.files.maxPartitionBytes","1g")
 
val df1 = spark.read.parquet("/home/xxx/data/tpcds_100G_parquet/customer").select("c_customer_sk").repartition(10000)
val df2 = df1.select($"c_customer_sk".alias("c_customer_sk2")).repartition(10000)
df1.join(df2, df1("c_customer_sk") > df2("c_customer_sk2"), "inner").count()
```

0.6:
```
spark.conf.set("spark.rapids.sql.exec.CartesianProductExec",false)
spark.conf.set("spark.rapids.sql.exec.BroadcastNestedLoopJoinExec",true)
spark.conf.set("spark.sql.autoBroadcastJoinThreshold",1000000000)
spark.conf.set("spark.sql.shuffle.partitions",200)
spark.conf.set("spark.sql.files.maxPartitionBytes","1g")

val df1 = spark.read.parquet("/home/xxx/data/tpcds_100G_parquet/customer").select("c_customer_sk").repartition(200)
val df2 = df1.select($"c_customer_sk".alias("c_customer_sk2")).repartition(200)
df1.join(df2, df1("c_customer_sk") > df2("c_customer_sk2"), "inner").count()
```

For 0.6, I realized the 10k partition is too huge so i reduced it to 200 and also reset batch size to default. 
In this way, it can maximize the GPU utilization.

I understand the 0.6 may have some throttle to avoid below error. 
So not sure if this is expected performance or not.
```
The resulting table has more rows than size_type's limit.
```

**Expected behavior**

0.6 and 0.5 have similar cross-join performance.

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

Spark 3.1.1 standalone cluster
Rapids accelerator 0.5 snapshot or 0.6 snapshot.

**Additional context**

**Comment by revans2** [comment]
This looks a bit like the issue fixed with https://github.com/NVIDIA/spark-rapids/pull/2356 `c_customer_sk` is a fixed width numeric column so the bug would cause the output batch size of the join to be 64 times smaller than it should be. It would be good to try with a newer version.

**Comment by viadea** [comment]
> This looks a bit like the issue fixed with #2356 `c_customer_sk` is a fixed width numeric column so the bug would cause the output batch size of the join to be 64 times smaller than it should be. It would be good to try with a newer version.

I built a latest 0.5 rapids with cuDF 0.20 latest one and the time is still 13mins.

**Comment by jlowe** [comment]
I was able to reproduce this on a smaller scale, using a SF=10 customer table, where using the setup for 0.5 resulted in 0.5 running in about 27 seconds on my desktop whereas 0.6 would take several minutes with the same configs.  I took an Nsight trace of each and here is a screenshot excerpt from each.
From 0.5:
![image](https://user-images.githubusercontent.com/1360766/118332742-2dd40900-b4d0-11eb-9913-663aad7e5686.png)
From 0.6:
![image](https://user-images.githubusercontent.com/1360766/118332757-39bfcb00-b4d0-11eb-8569-15d71334affb.png)

The core difference that results in such poor performance in 0.6 is because 0.6 is finally taking the `spark.rapids.sql.batchSizeBytes` setting seriously, working to ensure it does not produce a batch larger than the desired limit.  In 0.5 it would try to keep the join inputs below this value, but once the join occurred it would produce a batch far, far larger than this setting.  It's fast if you don't crash, not great if you do.

This behavior can be seen in the traces where a single input batch translates into a GPU acquisition followed by a single cross join, filter, hash aggregate, then GPU release.  In the 0.6 trace we can see that after a single GPU acquisition is a series of cross join, filter, and hash aggregate operations, showing that each input batch is being subdivided into smaller batches.  This is because it is trying to keep below the specified 1MB batch size, causing many more batches to occur.  All the extra batches leads to very poor GPU performance because of the tiny amounts of data involved.

What's very interesting is comparing the highly-tuned 0.5 query that barely works to this query in 0.6 with _no config tuning_ and using _only 12_ partitions:
```
scala> val df1 = spark.read.parquet("/home/jlowe/tpcds/10g/doubles/customer").select("c_customer_sk")
df1: org.apache.spark.sql.DataFrame = [c_customer_sk: int]                      

scala> val df2 = df1.select($"c_customer_sk".alias("c_customer_sk2"))
df2: org.apache.spark.sql.DataFrame = [c_customer_sk2: int]

scala> spark.time(df1.repartition(12).join(broadcast(df2), df1("c_customer_sk") > df2("c_customer_sk2"), "inner").count())
Time taken: 35575 ms
res0: Long = 124999750000
```

This is still a bit slower, but @revans2 pointed out that we are materializing the gather maps for this cross-join and could avoid doing this with a helper kernel in libcudf that can do partial cross-joins directly.  We'd expect this to save 80-100 microseconds per invocation which we'd expect to be somewhere between 10-16 seconds of savings overall on the wall-clock of that query, making it faster (and much more robust) than the 0.5 query.

Bottom line is that the poor performance comes from the batch size actually being honored on the join output starting with the out-of-core join support in 0.6.  Very small batch sizes will result in poor join performance due to the extreme internal slicing of the join.

**Comment by sameerz** [comment]
@viadea does the explanation above make sense, can we close this issue?

**Comment by viadea** [comment]
> @viadea does the explanation above make sense, can we close this issue?

Yes I reviewed @jlowe 's explanation and if this is due to the fact that we finally honored the batch size, then it seems a fix instead of performance regression. 
We can close this.
